refactor(linter): remove AstKind for SimpleAssignmentTarget

taearls opened the pull request:
Part of #11490

Progress Checklist:

- [x] Removed `SimpleAssignmentTarget` from the `ENUMS_WHITE_LIST`
- [x] regenerated the AST definitions
- [x] addressed the compile errors in `oxc_semantic`, `oxc_formatter`, `oxc_ast`, and `oxc_mangler`
- [x] all unit tests passing
- [x] all integration tests passing
- [x] just ready command runs successfully
- [x] fix conformance test failures (left comment for review)
- [x] self-review PR to see if I can simplify / minimize my changes 

Snapshot updates:
- After removing `SimpleAssignmentTarget`, the node ids in several snapshots were updated, which is expected since the generated node ids were changed.
- the prettier_conformance test snapshot for TS is slightly different (left a comment for feedback below)

taearls committed:
remove SimpleAssignmentTarget from ENUMS_WHITE_LIST

taearls committed:
regenerate ast, ast_visit, and formatter types

taearls committed:
fix non oxc_linter errors

taearls committed:
fix compile errors

taearls committed:
fix semantic tests

taearls committed:
get tests passing, but it's ugly

taearls committed:
simplify a bit

taearls committed:
get prefer_regexp_test tests to pass

taearls committed:
update snapshot for react_rules_of_hooks

taearls committed:
update snapshots

taearls committed:
update snapshot

taearls committed:
get tests to pass for func_names

taearls committed:
fix unit tests for id_length

overlookmotel commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -1791,8 +1791,28 @@ impl<'a> Visit<'a> for SemanticBuilder<'a> {
 
     fn visit_simple_assignment_target(&mut self, it: &SimpleAssignmentTarget<'a>) {
-        let kind = AstKind::SimpleAssignmentTarget(self.alloc(it));
-        self.enter_node(kind);
```
Yes, these lines need to be removed because `AstKind::SimpleAssignmentTarget` no longer exists.

But I think `self.leave_node(kind);` at bottom of the function also needs to be removed too for the same reason. `visit_identifier_reference` will already call `self.leave_node(AstKind::IdentifierReference)`, ditto for the other variants of `SimpleAssignmentTarget`.

The code as you have it now I think ends up calling `pop_ast_node` twice instead of once , so `current_node_id` gets out of sync with the traversal.

taearls committed:
remove extra line in builder.rs

taearls committed:
remove comments

taearls committed:
fix id_length rule unit tests

taearls committed:
fix no_import_assign tests

taearls committed:
get no_unused_private_class_members tests to pass

taearls committed:
remove debugging logs

taearls committed:
simplify code changes

overlookmotel commented:
I thought I saw a minute ago a diff in snapshots for a lint rule erroring on the `!!` in something like `if (x || !!(y || z))`. But now it's gone!

That looked like a pattern that the rule maybe *should* be catching - the `!!(...)` is unnecessary - so I thought you might have surfaced an unrelated bug. Can you remember what it was? We might want to fix that separately.

taearls committed:
get namespace tests to pass with minimal changes

taearls committed:
get prefer_jest_mocked tests passing

taearls committed:
get no_accumulating_spread tests to pass

taearls committed:
simplify code changes

taearls committed:
get only_used_in_recursion unit tests to pass with minimal changes

taearls committed:
get exhaustive_deps tests passing with minimal changes

taearls committed:
get more tests passing

taearls commented:
> I thought I saw a minute ago a diff in snapshots, a lint rule erroring on the `!!` in something like `if (x || !!(y || z))`. But now it's gone!
> 
> That looked like a pattern that the rule maybe _should_ be catching - the `!!(...)` is unnecessary - so I thought you might have surfaced an unrelated bug. Can you remember what it was exactly? We might want to fix that separately.



@overlookmotel I think you're talking about [this.](https://github.com/oxc-project/oxc/pull/12401/commits/d63d985afa780b3b3e4500726ace32d1f1e68414#diff-9cf418d8fa2bb364d41f5a736950b58d2da05ae76e3d7a7753674b9b207c63efL2127) 

The snapshot update was reverted once I had applied the change that you suggested the other day, which was to remove the `self.leave_node(kind)` from `visit_simple_assignment_target` in `builder.rs`.

My understanding was that the bug from my incomplete changes had caused the reporting in the snapshot to change, but it's very possible that I could be missing something here.

taearls committed:
get prefer_for_of tests to pass

taearls committed:
get no_accessor_recursion tests to pass

overlookmotel commented:
> @overlookmotel I think you're talking about [this.](https://github.com/oxc-project/oxc/pull/12401/commits/d63d985afa780b3b3e4500726ace32d1f1e68414#diff-9cf418d8fa2bb364d41f5a736950b58d2da05ae76e3d7a7753674b9b207c63efL2127)

Ah ha! The diff I saw was those same lines but all in green. But, as you've explained, that was reverting a previous change. So we're all good. Sorry, my question was a complete red herring.

taearls committed:
get prefer_array_find tests passing

taearls committed:
fix some lint warnings

taearls committed:
fix lint errors

taearls committed:
fix keep_names tests

taearls committed:
fix lint errors

taearls committed:
get all integration tests passing, update snapshots

taearls committed:
fix lint error

taearls committed:
get more conformance tests passing

taearls committed:
revert unnecessary change

taearls committed:
check for update expression in javascript checker

taearls committed:
get more conformance tests passing

taearls committed:
get more conformance tests passing

taearls committed:
add identifier reference to NeedsParentheses for TSTypeAssertion

overlookmotel commented:
> To that end, once I get everything working, I want to also take a few cycles to look for opportunities to simplify / minimize my changes. I'll likely leave comments on my PR on anything I'm not sure about.

Thank you for your very thorough approach. It's really appreciated.

FYI: What we've noticed while making other changes to `AstKind` is that the test coverage in linter is often not as good as it could be. Unfortunately, it's quite easy to introduce subtle bugs, and all tests may still pass. If you have time, it's ideal to go through the changes to linter and try to deduce the *logic* of the code, and make sure it's not changing, rather than relying solely on the tests.

This is our bad - other parts of Oxc have very thorough test coverage, which we can lean on when making changes, but the linter is less comprehensively tested. We might want to improve that (https://github.com/oxc-project/backlog/issues/170), but right now this is where we are.

However, figuring out the logic of each lint rule and trying to imagine all possible edge cases is both difficult and onerous, and tricky if you're not familiar with what the lint rules do. If you don't have time to delve that deep, please feel free to leave this until review.

Another side note: If you come across any unrelated bugs you want to fix, or edge cases that you want to add tests for, it'd be ideal to do that in separate PRs. This PR has quite a large diff, so it'd be preferable for ease of review if it's 100% refactor.

Hope this is useful background.

taearls committed:
fix(formatter): add assignment expression parentheses for TypeScript cast expressions

After SimpleAssignmentTarget removal from AstKind, TypeScript cast expressions
(TSAsExpression, TSSatisfiesExpression, TSTypeAssertion) were missing parentheses
when used as assignment targets.

Fixed by adding AssignmentExpression and AssignmentTargetWithDefault to the
parentheses logic for:
- ts_as_or_satisfies_needs_parens() (line 697)
- TSTypeAssertion needs_parentheses() (line 432)

This fixes formatting for cases like:
- (a as any) = value
- (<number>x) = value
- ({ a: (b as any) = 2000 } = x)

Result: TypeScript prettier conformance improved from 242/573 to 243/573

taearls committed:
update prettier ts snapshot

taearls committed:
undo unneeded changes in func_names rule

taearls committed:
simplify logic in id_length eslint rule

taearls committed:
simplify no_import_assign changes

taearls committed:
address lint error

taearls committed:
simplify logic in no_unused_private_class_members

taearls committed:
make simpler change in no_commonjs

taearls committed:
simplify changes in prefer_jest_mocked

taearls committed:
simplify changes in no_accumulating_spread, revert change to no_commonjs to prevent snapshot change

taearls committed:
simplify code change

taearls committed:
simplify prefer_for_of changes

taearls committed:
remove comment

taearls committed:
code cleanup in prefer_array_find

taearls committed:
revert unneeded changes

taearls committed:
revert unneeded changes in prefer_set_has

taearls committed:
remove comment

taearls committed:
fix doc comment

taearls committed:
return value directly instead of assigning to variable that is immediately returned

camc314 commented:
note to self, we need to fix all of the ecosystem CI failures before merging

https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/16538642569

(excluding oven-sh/bun as it's a diff issue)

camc314 committed:
fix private class false positive

camc314 commented:
Another false positive:

```
export class EdgelessColorPickerButton extends WithDisposable(LitElement) {
  readonly #select = (e: ColorEvent) => {
    e.stopPropagation();
    this.#pick(e.detail);
  };

  override render() {
    return html`
      <editor-menu-button
        .contentPadding=${this.tabContentPadding}
        .button=${html`
          <editor-icon-button
            aria-label=${this.label}
            .tooltip=${this.tooltip || this.label}
          >
            ${when(
              this.isText,
              () => html`
                <edgeless-text-color-icon
                  .color=${this.colorWithoutAlpha}
                ></edgeless-text-color-icon>
              `,
              () => html`
                <edgeless-color-button
                  .color=${this.colorWithoutAlpha}
                  .hollowCircle=${this.hollowCircle}
                ></edgeless-color-button>
              `
            )}
          </editor-icon-button>
        `}
      >
        ${choose(this.tabType, [
          [
            'normal',
            () => html`
              <div data-orientation="vertical">
                <slot name="other"></slot>
                <slot name="separator"></slot>
                <edgeless-color-panel
                  role="listbox"
                  class=${ifDefined(this.colorPanelClass)}
                  .value=${this.color}
                  .theme=${this.theme}
                  .palettes=${this.palettes}
                  .hollowCircle=${this.hollowCircle}
                  .hasTransparent=${false}
                  @select=${this.#select}
                >
                  ${when(
                    this.enableCustomColor,
                    () => html`
                      <edgeless-color-custom-button
                        slot="custom"
                        style=${styleMap(this.customButtonStyle)}
                        ?active=${this.isCustomColor}
                        @click=${this.switchToCustomTab}
                      ></edgeless-color-custom-button>
                    `
                  )}
                </edgeless-color-panel>
              </div>
            `,
          ],
          [
            'custom',
            () => {
              const packed = packColorsWith(
                this.theme,
                this.color,
                this.originalColor
              );
              const type = packed.type === 'palette' ? 'normal' : packed.type;
              const modes = packed.colors.map(
                preprocessColor(window.getComputedStyle(this))
              );

              return html`
                <edgeless-color-picker
                  class="custom"
                  .pick=${this.pick}
                  .colors=${{ type, modes }}
                ></edgeless-color-picker>
              `;
            },
          ],
        ])}
      </editor-menu-button>
    `;
  }

}
```

camc314 committed:
fix edge case

camc314 committed:
fix edge case

camc314 committed:
fix edge case

camc314 committed:
fix

camc314 committed:
remove comment
